contracts/core/script/admin/Admin.s.sol

```diff
@@ -137,4 +139,154 @@ contract Admin is Script {
+        vm.startBroadcast(deployer);
+        address impl = address(new Slashing());
+        vm.stopBroadcast();
+
+        _upgradeProxy(admin, Predeploys.Slashing, impl, data, false); // Slashing has no initializers
+
+        // TODO: add post upgrade tests
+    }
+
+    /**
+     * @notice Upgrade the OmniBridgeNative predeploy.
+     * @param admin     The address of the admin account, owner of the proxy admin
+     * @param deployer  The address of the account that will deploy the new implementation.
+     */
+    function upgradeBridgeNative(address admin, address deployer, bytes calldata data) public {
+        vm.startBroadcast(deployer);
+        address impl = address(new OmniBridgeNative());
+        vm.stopBroadcast();
+
+        _upgradeProxy(admin, Predeploys.OmniBridgeNative, impl, data);
+
+        // TODO: add post upgrade tests
+    }
+
+    /**
+     * @notice Upgrade the OmniBridgeL1 contract.
+     * @param admin     The address of the admin account, owner of the proxy admin
+     * @param deployer  The address of the account that will deploy the new implementation.
+     * @param proxy     The address of the proxy to upgrade.
+     */
+    function upgradeBridgeL1(address admin, address deployer, address proxy, bytes calldata data) public {
+        address token = address(OmniBridgeL1(proxy).token());
+
+        vm.startBroadcast(deployer);
+        address impl = address(new OmniBridgeL1(token));
+        vm.stopBroadcast();
+
+        _upgradeProxy(admin, proxy, impl, data);
+
+        // TODO: add post upgrade tests
+    }
+
+    /**
+     * @notice Upgrade the PortalRegistry predeploy.
+     * @param admin     The address of the admin account, owner of the proxy admin
+     * @param deployer  The address of the account that will deploy the new implementation.
+     */
+    function upgradePortalRegistry(address admin, address deployer, bytes calldata data) public {
+        vm.startBroadcast(deployer);
+        address impl = address(new PortalRegistry());
+        vm.stopBroadcast();
+
+        _upgradeProxy(admin, Predeploys.PortalRegistry, impl, data);
+
+        // TODO: add post upgrade tests
+    }
+
+    /**
+     * @notice Upgrade a proxy contract.
+     * @param admin     The address of the admin account, owner of the proxy admin
+     * @param proxy     The address of the proxy to upgrade.
+     * @param impl      The address of the new implementation.
+     * @param data      Calldata to execute after upgrading the contract.
+     */
+    function _upgradeProxy(address admin, address proxy, address impl, bytes calldata data) internal {
+        _upgradeProxy(admin, proxy, impl, data, true);
+    }
+
+    function _upgradeProxy(address admin, address proxy, address impl, bytes calldata data, bool initializable)
+        internal
+    {
+        vm.startBroadcast(admin);
+        address proxyAdmin = EIP1967Helper.getAdmin(proxy);
```

Comment:
can this be outside of the broadcast?

Reply:
yes good point. in this case it does not matter but better to be careful about what's in broadcast